crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -712,10 +708,65 @@ impl IndexerReader {
                 )
                 // we filter the `tx_global_order` table because it is indexed by digest,
                 // optimistic_transactions table is not
-                .filter(tx_global_order::tx_digest.eq_any(missing_digests))
+                .filter(tx_global_order::tx_digest.eq_any(digests))
                 .select(OptimisticTransaction::as_select())
                 .load::<OptimisticTransaction>(conn)
-        })?;
+        })
+    }
+
+    async fn get_single_transaction(
+        &self,
+        digest: TransactionDigest,
+    ) -> IndexerResult<Option<StoredTransaction>> {
+        let digest_bytes = digest.inner().to_vec();
+        let checkpointed_tx_future = {
+            let this = self.clone();
+            let digest_bytes = digest_bytes.clone();
+            tokio::task::spawn_blocking(move || {
+                this.get_checkpointed_transactions(&[digest_bytes])
+                    .map(|mut txs| txs.pop())
+            })
+        };
+        let optimistic_tx_future = {
+            let this = self.clone();
+            tokio::task::spawn_blocking(move || {
+                this.get_optimistic_transactions(&[digest_bytes])
+                    .map(|mut txs| txs.pop().map(Into::into))
+            })
+        };
+
+        let mut tx_futures = FuturesUnordered::new();
+        tx_futures.push(checkpointed_tx_future);
+        tx_futures.push(optimistic_tx_future);
+
+        while let Some(result) = tx_futures.next().await {
+            if let Some(tx) = result?? {
+                return Ok(Some(tx));
+            }
+        }
+
+        Ok(None)
+    }
+
+    fn multi_get_transactions(
+        &self,
+        digests: &[TransactionDigest],
+    ) -> Result<Vec<StoredTransaction>, IndexerError> {
+        let digests: Vec<Vec<u8>> = digests.iter().map(|d| d.inner().to_vec()).collect();
+        let checkpointed_txs = self.get_checkpointed_transactions(&digests)?;
+
+        if checkpointed_txs.len() == digests.len() {
+            return Ok(checkpointed_txs);
+        }
+
+        let mut missing_digests: HashSet<Vec<u8>> = digests.iter().cloned().collect();
+        for tx in &checkpointed_txs {
+            missing_digests.remove(&tx.transaction_digest);
+        }
+        let missing_digests = missing_digests.into_iter().collect::<Vec<_>>();
```

Comment:
Currently, we're cloning all digests upfront into a HashSet, then removing matching digests. The suggested approach aims to only clone the missing digests instead of cloning all digests upfront.

```suggestion
        let checkpointed_tx_digests_set = checkpointed_txs
            .iter()
            .map(|tx| &tx.transaction_digest)
            .collect::<HashSet<&Vec<u8>>>();

        let missing_digests = digests
            .iter()
            .filter(|digest| !checkpointed_tx_digests_set.contains(digest))
            .cloned()
            .collect::<Vec<Vec<u8>>>();
```